In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

In [2]:
# set the paths
DATA_DIR = Path("../data/raw")

calendar_path = DATA_DIR / "calendar.csv"
prices_path = DATA_DIR / "sell_prices.csv"
sales_path = DATA_DIR / "sales_train_evaluation.csv"

In [3]:
#Load the datasets
calendar = pd.read_csv(calendar_path)
prices = pd.read_csv(prices_path)
sales = pd.read_csv(sales_path)

In [4]:
print("Calendar shape:", calendar.shape)
print("Prices shape:", prices.shape)
print("Sales shape:", sales.shape)

Calendar shape: (1969, 14)
Prices shape: (6841121, 4)
Sales shape: (30490, 1947)


In [5]:
display(calendar.head())
display(prices.head())
display(sales.iloc[:5, :12])

,date,wm_yr_wk,weekday,wday,month,year,d,event_name_1,event_type_1,event_name_2,event_type_2,snap_CA,snap_TX,snap_WI
0,2011-01-29,11101,Saturday,1,1,2011,d_1,NaN,NaN,NaN,NaN,0,0,0
1,2011-01-30,11101,Sunday,2,1,2011,d_2,NaN,NaN,NaN,NaN,0,0,0
2,2011-01-31,11101,Monday,3,1,2011,d_3,NaN,NaN,NaN,NaN,0,0,0
3,2011-02-01,11101,Tuesday,4,2,2011,d_4,NaN,NaN,NaN,NaN,1,1,0
4,2011-02-02,11101,Wednesday,5,2,2011,d_5,NaN,NaN,NaN,NaN,1,0,1


,store_id,item_id,wm_yr_wk,sell_price
0,CA_1,HOBBIES_1_001,11325,9.58
1,CA_1,HOBBIES_1_001,11326,9.58
2,CA_1,HOBBIES_1_001,11327,8.26
3,CA_1,HOBBIES_1_001,11328,8.26
4,CA_1,HOBBIES_1_001,11329,8.26


,id,item_id,dept_id,cat_id,store_id,state_id,d_1,d_2,d_3,d_4,d_5,d_6
0,HOBBIES_1_001_CA_1_evaluation,HOBBIES_1_001,HOBBIES_1,HOBBIES,CA_1,CA,0,0,0,0,0,0
1,HOBBIES_1_002_CA_1_evaluation,HOBBIES_1_002,HOBBIES_1,HOBBIES,CA_1,CA,0,0,0,0,0,0
2,HOBBIES_1_003_CA_1_evaluation,HOBBIES_1_003,HOBBIES_1,HOBBIES,CA_1,CA,0,0,0,0,0,0
3,HOBBIES_1_004_CA_1_evaluation,HOBBIES_1_004,HOBBIES_1,HOBBIES,CA_1,CA,0,0,0,0,0,0
4,HOBBIES_1_005_CA_1_evaluation,HOBBIES_1_005,HOBBIES_1,HOBBIES,CA_1,CA,0,0,0,0,0,0


In [ ]:
# column names
print("Calendar columns:")
print(calendar.columns.tolist())

print("\nPrice columns:")
print(prices.columns.tolist())

print("\nSales identifier columns:")
print(sales.columns[:6].tolist())

print("\nNumber of daily sales columns:")
print(len(sales.columns[6:]))

print("\nFirst 5 daily columns:")
print(sales.columns[6:11].tolist())

print("\nLast 5 daily columns:")
print(sales.columns[-5:].tolist())

Calendar columns:
['date', 'wm_yr_wk', 'weekday', 'wday', 'month', 'year', 'd', 'event_name_1', 'event_type_1', 'event_name_2', 'event_type_2', 'snap_CA', 'snap_TX', 'snap_WI']

Price columns:
['store_id', 'item_id', 'wm_yr_wk', 'sell_price']

Sales identifier columns:
['id', 'item_id', 'dept_id', 'cat_id', 'store_id', 'state_id']

Number of daily sales columns:
1941

First 5 daily columns:
['d_1', 'd_2', 'd_3', 'd_4', 'd_5']

Last 5 daily columns:
['d_1937', 'd_1938', 'd_1939', 'd_1940', 'd_1941']


In [7]:
# Check data types and memory usage of each dataset

for name, df in {
    "Calendar": calendar,
    "Prices": prices,
    "Sales": sales
}.items():
    memory_mb = df.memory_usage(deep=True).sum() / 1024**2

    print(f"\n{name}")
    print("-" * 40)
    print(df.dtypes.value_counts())
    print(f"Memory usage: {memory_mb:.2f} MB")


Calendar
----------------------------------------
str      7
int64    7
Name: count, dtype: int64
Memory usage: 0.67 MB

Prices
----------------------------------------
str        2
int64      1
float64    1
Name: count, dtype: int64
Memory usage: 853.13 MB

Sales
----------------------------------------
int64    1941
str         6
Name: count, dtype: int64
Memory usage: 461.90 MB


In [8]:
print("Calendar:")
display(calendar[["date", "wm_yr_wk", "weekday", "month", "year", "d"]].dtypes)

print("\nPrices:")
display(prices.dtypes)

print("\nSales identifiers:")
display(sales.iloc[:, :6].dtypes)

print("\nSales demand columns:")
display(sales.iloc[:, 6:11].dtypes)

Calendar:


date          str
wm_yr_wk    int64
weekday       str
month       int64
year        int64
d             str
dtype: object


Prices:


store_id          str
item_id           str
wm_yr_wk        int64
sell_price    float64
dtype: object


Sales identifiers:


id          str
item_id     str
dept_id     str
cat_id      str
store_id    str
state_id    str
dtype: object


Sales demand columns:


d_1    int64
d_2    int64
d_3    int64
d_4    int64
d_5    int64
dtype: object

In [9]:
print("Missing values")
print("-" * 40)

print("\nCalendar:")
print(calendar.isnull().sum())

print("\nPrices:")
print(prices.isnull().sum())

print("\nSales:")
print("Total missing values:", sales.isnull().sum().sum())

Missing values
----------------------------------------

Calendar:
date               0
wm_yr_wk           0
weekday            0
wday               0
month              0
year               0
d                  0
event_name_1    1807
event_type_1    1807
event_name_2    1964
event_type_2    1964
snap_CA            0
snap_TX            0
snap_WI            0
dtype: int64

Prices:
store_id      0
item_id       0
wm_yr_wk      0
sell_price    0
dtype: int64

Sales:
Total missing values: 0


In [10]:
print("Duplicate rows")
print("-" * 40)
print("Calendar:", calendar.duplicated().sum())
print("Prices:", prices.duplicated().sum())
print("Sales:", sales.duplicated().sum())

print("\nDataset coverage")
print("-" * 40)
print("Unique items:", sales["item_id"].nunique())
print("Unique departments:", sales["dept_id"].nunique())
print("Unique categories:", sales["cat_id"].nunique())
print("Unique stores:", sales["store_id"].nunique())
print("Unique states:", sales["state_id"].nunique())

print("\nPrice range")
print("-" * 40)
print(prices["sell_price"].describe())

Duplicate rows
----------------------------------------
Calendar: 0
Prices: 0
Sales: 0

Dataset coverage
----------------------------------------
Unique items: 3049
Unique departments: 7
Unique categories: 3
Unique stores: 10
Unique states: 3

Price range
----------------------------------------
count    6.841121e+06
mean     4.410952e+00
std      3.408814e+00
min      1.000000e-02
25%      2.180000e+00
50%      3.470000e+00
75%      5.840000e+00
max      1.073200e+02
Name: sell_price, dtype: float64


In [11]:
# Explore the product and store hierarchy in the sales data

print("Categories:")
print(sales["cat_id"].value_counts())

print("\nDepartments:")
print(sales["dept_id"].value_counts())

print("\nStores:")
print(sales["store_id"].value_counts())

print("\nStates:")
print(sales["state_id"].value_counts())

Categories:
cat_id
FOODS        14370
HOUSEHOLD    10470
HOBBIES       5650
Name: count, dtype: int64

Departments:
dept_id
FOODS_3        8230
HOUSEHOLD_1    5320
HOUSEHOLD_2    5150
HOBBIES_1      4160
FOODS_2        3980
FOODS_1        2160
HOBBIES_2      1490
Name: count, dtype: int64

Stores:
store_id
CA_1    3049
CA_2    3049
CA_3    3049
CA_4    3049
TX_1    3049
TX_2    3049
TX_3    3049
WI_1    3049
WI_2    3049
WI_3    3049
Name: count, dtype: int64

States:
state_id
CA    12196
TX     9147
WI     9147
Name: count, dtype: int64


In [12]:
# Measure how much historical price variation exists for each product-store combination

price_variation = (
    prices.groupby(["store_id", "item_id"])["sell_price"]
    .agg(["count", "nunique", "min", "max", "mean"])
    .reset_index()
)

price_variation["price_range"] = (
    price_variation["max"] - price_variation["min"]
)

display(price_variation.head())

print("\nProduct-store combinations:", len(price_variation))
print(
    "Combinations with multiple prices:",
    (price_variation["nunique"] > 1).sum()
)

print(
    "Percentage with multiple prices:",
    round((price_variation["nunique"] > 1).mean() * 100, 2),
    "%"
)

,store_id,item_id,count,nunique,min,max,mean,price_range
0,CA_1,FOODS_1_001,282,2,2.00,2.24,2.169362,0.24
1,CA_1,FOODS_1_002,282,3,7.88,9.48,8.937447,1.60
2,CA_1,FOODS_1_003,282,2,2.88,3.23,2.976809,0.35
3,CA_1,FOODS_1_004,225,3,1.50,1.96,1.851378,0.46
4,CA_1,FOODS_1_005,282,3,2.94,3.54,3.334468,0.60



Product-store combinations: 30490
Combinations with multiple prices: 22243
Percentage with multiple prices: 72.95 %


In [13]:
# Summarize the number of unique prices and price ranges across products

display(
    price_variation[["nunique", "price_range"]]
    .describe()
)

display(
    price_variation
    .sort_values("nunique", ascending=False)
    .head(10)
)

,nunique,price_range
count,30490.000000,30490.000000
mean,2.769367,0.612636
std,1.893627,1.303884
min,1.000000,0.000000
25%,1.000000,0.000000
50%,2.000000,0.280000
75%,4.000000,0.720000
max,21.000000,104.060000


,store_id,item_id,count,nunique,min,max,mean,price_range
18387,TX_3,FOODS_1_095,282,21,3.44,4.62,4.111702,1.18
12289,TX_1,FOODS_1_095,282,21,3.44,4.62,4.116950,1.18
21469,WI_1,FOODS_1_129,282,20,5.04,9.00,7.994610,3.96
3142,CA_2,FOODS_1_095,282,18,2.66,3.42,3.026312,0.76
27567,WI_3,FOODS_1_129,282,18,5.04,9.00,8.031631,3.96
15338,TX_2,FOODS_1_095,282,18,3.44,4.62,4.103050,1.18
9240,CA_4,FOODS_1_095,282,18,2.66,3.42,3.034504,0.76
93,CA_1,FOODS_1_095,282,18,2.66,3.42,3.027482,0.76
24518,WI_2,FOODS_1_129,282,18,7.02,9.00,8.008262,1.98
27535,WI_3,FOODS_1_096,282,17,7.02,9.00,7.999645,1.98


In [14]:
# Summarize daily demand values across the complete sales history

demand_data = sales.iloc[:, 6:]

print("Total units sold:", demand_data.to_numpy().sum())
print("Average daily demand:", round(demand_data.to_numpy().mean(), 2))
print("Maximum daily demand:", demand_data.to_numpy().max())

zero_percentage = (demand_data.to_numpy() == 0).mean() * 100
print("Zero-demand observations:", round(zero_percentage, 2), "%")

Total units sold: 66927173
Average daily demand: 1.13
Maximum daily demand: 763
Zero-demand observations: 68.0 %


In [15]:
# Compare overall demand levels across categories and stores

sales_summary = sales.iloc[:, :6].copy()
sales_summary["total_demand"] = demand_data.sum(axis=1)
sales_summary["avg_daily_demand"] = demand_data.mean(axis=1)

print("Demand by category:")
display(
    sales_summary.groupby("cat_id")["total_demand"]
    .sum()
    .sort_values(ascending=False)
)

print("Demand by store:")
display(
    sales_summary.groupby("store_id")["total_demand"]
    .sum()
    .sort_values(ascending=False)
)

Demand by category:


cat_id
FOODS        45922427
HOUSEHOLD    14764090
HOBBIES       6240656
Name: total_demand, dtype: int64

Demand by store:


store_id
CA_3    11363540
CA_1     7832248
TX_2     7329642
WI_2     6697988
WI_3     6542557
TX_3     6205940
CA_2     5818395
TX_1     5692823
WI_1     5261506
CA_4     4182534
Name: total_demand, dtype: int64

In [16]:
# Verify the common keys used to connect sales, prices, and calendar data

sales_items = set(sales["item_id"].unique())
price_items = set(prices["item_id"].unique())

sales_stores = set(sales["store_id"].unique())
price_stores = set(prices["store_id"].unique())

sales_days = set(sales.columns[6:])
calendar_days = set(calendar["d"])

print("Sales items found in prices:", sales_items.issubset(price_items))
print("Sales stores found in prices:", sales_stores.issubset(price_stores))
print("Sales days found in calendar:", sales_days.issubset(calendar_days))

print("\nUnique sales items:", len(sales_items))
print("Unique price items:", len(price_items))

print("Unique sales stores:", len(sales_stores))
print("Unique price stores:", len(price_stores))

Sales items found in prices: True
Sales stores found in prices: True
Sales days found in calendar: True

Unique sales items: 3049
Unique price items: 3049
Unique sales stores: 10
Unique price stores: 10


## Data Understanding Summary

- The dataset contains 3,049 products across 10 stores and 3 states.
- Products belong to 7 departments and 3 main categories: FOODS, HOUSEHOLD, and HOBBIES.
- Daily sales history contains 1,941 days of demand observations.
- The price dataset contains more than 6.8 million weekly price records.
- 72.95% of product-store combinations have experienced more than one historical selling price.
- Some product-store combinations contain substantial historical price variation, making the dataset suitable for price-sensitivity analysis.
- Sales and price datasets contain no missing values or duplicate rows.
- Missing values in calendar event columns mainly represent dates without special events.
- The three datasets can be connected using product/store identifiers, calendar day identifiers, and week identifiers.
- Because converting the complete sales dataset to long format would create roughly 59 million rows, memory-efficient processing will be important in later stages.